mne/decoding/gsoc.py

```diff
@@ -0,0 +1,216 @@
+            Data of epochs.
+        y : ndarray shape(n_samples,)
+            Target values.
+
+        Returns
+        -------
+        self : XdawnTransformer instance
+            The XdawnTransformer instance.
+        """
+        from sklearn.preprocessing import LabelEncoder
+
+        if X.ndim != 2 or not isinstance(X, np.ndarray):
+            raise ValueError("X should be 2 dimensional ndarray")
+        if not isinstance(y, np.ndarray):
+            raise ValueError("Labels must be numpy array")
+
+        epochs_data = X.reshape(X.shape[0], self.n_chan, X.shape[1] //
+                                self.n_chan)
+        # Extract signal covariance
+        self._get_signal_cov(epochs_data)
+
+        # estimates evoked covariance
+        evokeds = dict()
+        toeplitz = dict()
+        classes = LabelEncoder().fit(y).classes_
+        shape = epochs_data.shape
+        for eid in classes:
+            mean_data = np.mean(epochs_data[eid].reshape(-1, shape[1],
+                                shape[2]), axis=0)
+            evokeds[eid] = mean_data
+            toeplitz[eid] = 1.0
+
+        for eid in classes:
+            data = np.dot(evokeds[eid], toeplitz[eid])
+            self.evokeds_cov_[eid] = _regularized_covariance(data, self.reg)
+
+        # estimates spatial filters
+        for eid in classes:
+            self._fit_xdawn(eid)
+
+        # store some values
+        self.event_id = classes
+        self.exclude = list(range(self.n_components, self.n_chan))
+        return self
+
+    def transform(self, X):
+        """Apply Xdawn dim reduction.
+
+        Parameters
+        ----------
+        X : ndarray, shape(n_channels, n_times * n_freq)
+            data of epochs.
+
+        Returns
+        -------
+        X : ndarray, shape (n_epochs, n_components * event_types * n_times)
+            Spatially filtered signals.
+        """
+        if isinstance(X, np.ndarray):
+            data = X
+            shape = X.shape
+            data = X.reshape(X.shape[0], self.n_chan, X.shape[1] /
+                             self.n_chan)
+
+        else:
+            raise ValueError('Data input must be of type numpy array')
+
+        # create full matrix of spatial filter
+        result = self._transform_xdawn(data)
+        shape = result.shape
+        return result.reshape(-1, shape[1] * shape[2])
+
+    def fit_transform(self, X, y):
+        """First fit the data, then transform
+
+        Parameters
+        ----------
+        X : ndarray, shape(n_channels, n_times * n_freq)
+            data of epochs.
+        y : ndarray, shape(n_samples,)
+            labels of data.
+
+        Returns
+        -------
+        X : ndarray, shape(n_epochs, n_components * event_types * n_times)
+            spatially filtered signals.
+        """
+        self.fit(X, y)
+        return self.transform(X)
+
+    def inverse_transform(self, X, event_id=None, include=None, exclude=None):
```

Comment:
XDawnTransformer won't cover the overlapping events, so event_id isn't necessary right?

include and exclude should be set be `n_component`